File: src/commands/tag-stats.ts
import { readFile } from 'fs/promises';
import type { Command } from 'commander';
import { join } from 'path';
import { existsSync } from 'fs';
import chalk from 'chalk';
import { glob } from 'tinyglobby';
import * as Gherkin from '@cucumber/gherkin';
import * as Messages from '@cucumber/messages';
import type { Tags } from '../types/tags';

interface TagCount {
  tag: string;
  count: number;
}

interface CategoryStats {
  name: string;
  tags: TagCount[];
}

interface TagStatsOptions {
  cwd?: string;
}

interface TagStatsResult {
  success: boolean;
  totalFiles: number;
  uniqueTags: number;
  totalOccurrences: number;
  categories: CategoryStats[];
  unusedTags: string[];
  tagsFileFound: boolean;
  invalidFiles: string[];
}

export async function tagStats(
  options: TagStatsOptions = {}
): Promise<TagStatsResult> {
  const cwd = options.cwd || process.cwd();

  // Load tag registry from tags.json (optional)
  let tagsData: Tags | null = null;
  let tagsFileFound = true;
  try {
    tagsData = await loadTagsJson(cwd);
  } catch {
    tagsFileFound = false;
  }

  // Get all feature files
  const files = await glob(['spec/features/**/*.feature'], {
    cwd,
    absolute: false,
  });

  if (files.length === 0) {
    return {
      success: true,
      totalFiles: 0,
      uniqueTags: 0,
      totalOccurrences: 0,
      categories: [],
      unusedTags: [],
      tagsFileFound,
      invalidFiles: [],
    };
  }

  // Extract tags from all feature files
  const tagCounts = new Map<string, number>();
  const invalidFiles: string[] = [];

  for (const file of files) {
    try {
      const content = await readFile(join(cwd, file), 'utf-8');

      const uuidFn = Messages.IdGenerator.uuid();
      const builder = new Gherkin.AstBuilder(uuidFn);
      const matcher = new Gherkin.GherkinClassicTokenMatcher();
      const parser = new Gherkin.Parser(builder, matcher);

      let gherkinDocument;
      try {
        gherkinDocument = parser.parse(content);
      } catch {
        invalidFiles.push(file);
        continue;
      }

      if (!gherkinDocument.feature) {
        continue;
      }

      // Count tags from this feature
      const tags = gherkinDocument.feature.tags.map(t => t.name);
      for (const tag of tags) {
        tagCounts.set(tag, (tagCounts.get(tag) || 0) + 1);
      }
    } catch {
      invalidFiles.push(file);
    }
  }

  const uniqueTags = tagCounts.size;
  const totalOccurrences = Array.from(tagCounts.values()).reduce(
    (sum, count) => sum + count,
    0
  );

  // Build statistics by category
  const categories: CategoryStats[] = [];
  const registeredTags = new Set<string>();

  if (tagsData) {
    // Group by categories from tags.json
    for (const category of tagsData.categories) {
      const tagStats: TagCount[] = [];

      for (const tagDef of category.tags) {
        const tag = tagDef.name;
        registeredTags.add(tag);
        const count = tagCounts.get(tag) || 0;
        if (count > 0) {
          tagStats.push({ tag, count });
        }
      }

      // Sort by count descending
      tagStats.sort((a, b) => b.count - a.count);

      if (tagStats.length > 0) {
        categories.push({ name: category.name, tags: tagStats });
      }
    }

    // Find unregistered tags
    const unregisteredTags: TagCount[] = [];
    for (const [tag, count] of tagCounts.entries()) {
      if (!registeredTags.has(tag)) {
        unregisteredTags.push({ tag, count });
      }
    }

    if (unregisteredTags.length > 0) {
      unregisteredTags.sort((a, b) => b.count - a.count);
      categories.push({ name: 'Unregistered', tags: unregisteredTags });
    }
  } else {
    // No tags.json - all tags are unregistered
    const allTags: TagCount[] = Array.from(tagCounts.entries()).map(
      ([tag, count]) => ({
        tag,
        count,
      })
    );
    allTags.sort((a, b) => b.count - a.count);
    categories.push({ name: 'Unregistered', tags: allTags });
  }

  // Find unused registered tags
  const unusedTags: string[] = [];
  if (tagsData) {
    for (const category of tagsData.categories) {
      for (const tagDef of category.tags) {
        const tag = tagDef.name;
        registeredTags.add(tag);
        if (!tagCounts.has(tag)) {
          unusedTags.push(tag);
        }
      }
    }
    unusedTags.sort();
  }

  return {
    success: true,
    totalFiles: files.length,
    uniqueTags,
    totalOccurrences,
    categories,
    unusedTags,
    tagsFileFound,
    invalidFiles,
  };
}

export async function tagStatsCommand(): Promise<void> {
  try {
    const result = await tagStats();

    // Display overall statistics
    console.log(chalk.bold.blue('\nTag Usage Statistics'));
    console.log(chalk.gray('─'.repeat(50)));
    console.log(`Total feature files: ${chalk.cyan(result.totalFiles)}`);
    console.log(`Unique tags used: ${chalk.cyan(result.uniqueTags)}`);
    console.log(
      `Total tag occurrences: ${chalk.cyan(result.totalOccurrences)}`
    );

    if (!result.tagsFileFound) {
      console.log(chalk.yellow('\n⚠ Warning: spec/tags.json not found'));
    }

    if (result.invalidFiles.length > 0) {
      console.log(
        chalk.yellow(
          `\n⚠ Warning: ${result.invalidFiles.length} file(s) with invalid syntax skipped:`
        )
      );
      for (const file of result.invalidFiles) {
        console.log(chalk.yellow(`  - ${file}`));
      }
    }

    // Display per-category statistics
    if (result.categories.length > 0) {
      console.log(chalk.bold.blue('\n\nTag Counts by Category'));
      console.log(chalk.gray('─'.repeat(50)));

      for (const category of result.categories) {
        console.log(
          chalk.bold(`\n${category.name}`) +
            chalk.gray(` (${category.tags.length} tags)`)
        );
        for (const tagCount of category.tags) {
          console.log(
            `  ${chalk.green(tagCount.tag.padEnd(30))} ${chalk.cyan(tagCount.count)}`
          );
        }
      }
    }

    // Display unused tags
    if (result.unusedTags.length > 0) {
      console.log(chalk.bold.blue('\n\nUnused Registered Tags'));
      console.log(chalk.gray('─'.repeat(50)));
      console.log(
        chalk.yellow(
          `${result.unusedTags.length} registered tag(s) not used in any feature file:\n`
        )
      );
      for (const tag of result.unusedTags) {
        console.log(chalk.yellow(`  ${tag}`));
      }
    }

    console.log('');
    process.exit(0);
  } catch (error: any) {
    console.error(chalk.red('Error:'), error.message);
    process.exit(2);
  }
}

async function loadTagsJson(cwd: string): Promise<Tags> {
  const tagsJsonPath = join(cwd, 'spec', 'tags.json');

  if (!existsSync(tagsJsonPath)) {
    throw new Error('tags.json not found: spec/tags.json');
  }

  const content = await readFile(tagsJsonPath, 'utf-8');
  const tagsData: Tags = JSON.parse(content);

  return tagsData;
}

export function registerTagStatsCommand(program: Command): void {
  program
    .command('tag-stats')
    .description('Show tag usage statistics across all feature files')
    .action(tagStatsCommand);
}

File: codelet/fspec-core/src/commands/tag_stats.rs
//! `tag-stats` — Rust port of `src/commands/tag-stats.ts` (RPC-310).
//!
//! Aggregates feature-level tag usage across every `*.feature` file under
//! `spec/features/`, projects the counts into the categories declared by
//! the project tag registry (when present), and emits either a pretty-
//! printed JSON payload (dispatcher path) or a plain-text summary
//! (CLI path).
//!
//! Both invocation paths (the LLM-facing dispatcher AND the standalone
//! fspec Rust binary's clap subcommand) call this single function —
//! RPC-003 §7/§11 two-front-doors invariant.
//!
//! ## TS parity rules
//!
//! * **No auto-create** of either the tag registry or `spec/features/`.
//!   Missing OR malformed registry → `tagsFileFound=false`.
//! * **Feature-level tags only**: parity with the TS
//!   `gherkinDocument.feature.tags` projection. Scenario, scenario-outline,
//!   background and rule-nested tags are excluded.
//! * **Invalid Gherkin → invalidFiles**: a feature file whose contents
//!   contain no `Feature:` keyword is recorded under `invalidFiles` and
//!   contributes no counts.
//! * **Categories preserve declaration order**: the registry array on
//!   disk is the render order. Tags WITHIN each category sort by count
//!   descending (stable for ties).
//! * **Unregistered bucket**: tags observed in feature files that do
//!   not appear in any registered category collect under a synthetic
//!   `Unregistered` category appended after the registered ones, sorted
//!   desc-by-count.
//! * **unusedTags**: registered tags with zero occurrences,
//!   alphabetically sorted. Only populated when the registry parsed.
//! * **JSON field order**: `success, totalFiles, uniqueTags,
//!   totalOccurrences, categories, unusedTags, tagsFileFound,
//!   invalidFiles` (declaration order, NOT alphabetical).

use std::path::Path;

use serde::{Deserialize, Serialize};

use crate::error::FspecCoreError;
use crate::io::feature_glob::glob_feature_files;
use crate::types::tags::TagsData;

#[derive(Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
struct TagStatsArgs {
    #[serde(default)]
    format: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct TagStatsResult {
    success: bool,
    total_files: usize,
    unique_tags: usize,
    total_occurrences: usize,
    categories: Vec<CategoryStats>,
    unused_tags: Vec<String>,
    tags_file_found: bool,
    invalid_files: Vec<String>,
}

#[derive(Debug, Serialize)]
struct CategoryStats {
    name: String,
    tags: Vec<TagCount>,
}

#[derive(Debug, Serialize)]
struct TagCount {
    tag: String,
    count: usize,
}

pub async fn run(args_json: &str, project_root: &Path) -> Result<String, FspecCoreError> {
    let args: TagStatsArgs =
        serde_json::from_str(args_json).map_err(|e| FspecCoreError::InvalidArgs {
            command: "tag-stats",
            reason: format!("failed to parse args: {e}"),
        })?;

    let result = compute_stats(project_root);

    match args.format.as_deref() {
        Some("text") => Ok(render_text(&result)),
        _ => serde_json::to_string_pretty(&result).map_err(|e| FspecCoreError::InvalidArgs {
            command: "tag-stats",
            reason: format!("failed to serialize result: {e}"),
        }),
    }
}

fn compute_stats(project_root: &Path) -> TagStatsResult {
    let (tags_data, tags_file_found) = load_tags_registry(project_root);

    let files = match glob_feature_files(project_root) {
        Ok(v) => v,
        Err(FspecCoreError::DirectoryNotFound { .. }) => Vec::new(),
        Err(_) => Vec::new(),
    };

    if files.is_empty() {
        return TagStatsResult {
            success: true,
            total_files: 0,
            unique_tags: 0,
            total_occurrences: 0,
            categories: Vec::new(),
            unused_tags: Vec::new(),
            tags_file_found,
            invalid_files: Vec::new(),
        };
    }

    let mut tag_counts: Vec<(String, usize)> = Vec::new();
    let mut invalid_files: Vec<String> = Vec::new();

    for rel in &files {
        let abs = project_root.join(rel);
        let content = match std::fs::read_to_string(&abs) {
            Ok(s) => s,
            Err(_) => {
                invalid_files.push(rel.clone());
                continue;
            }
        };

        match parse_feature_tags(&content) {
            Some(tags) => {
                for t in tags {
                    bump_count(&mut tag_counts, &t);
                }
            }
            None => {
                invalid_files.push(rel.clone());
            }
        }
    }

    let unique_tags = tag_counts.len();
    let total_occurrences: usize = tag_counts.iter().map(|(_, c)| *c).sum();

    let (categories, unused_tags) =
        build_categories(&tag_counts, tags_data.as_ref(), tags_file_found);

    TagStatsResult {
        success: true,
        total_files: files.len(),
        unique_tags,
        total_occurrences,
        categories,
        unused_tags,
        tags_file_found,
        invalid_files,
    }
}

fn load_tags_registry(project_root: &Path) -> (Option<TagsData>, bool) {
    let path = project_root.join("spec").join("tags.json");
    let raw = match std::fs::read_to_string(&path) {
        Ok(s) => s,
        Err(_) => return (None, false),
    };
    match serde_json::from_str::<TagsData>(&raw) {
        Ok(d) => (Some(d), true),
        Err(_) => (None, false),
    }
}

fn bump_count(counts: &mut Vec<(String, usize)>, tag: &str) {
    if let Some(entry) = counts.iter_mut().find(|(k, _)| k == tag) {
        entry.1 += 1;
    } else {
        counts.push((tag.to_string(), 1));
    }
}

fn build_categories(
    tag_counts: &[(String, usize)],
    tags_data: Option<&TagsData>,
    tags_file_found: bool,
) -> (Vec<CategoryStats>, Vec<String>) {
    let mut categories: Vec<CategoryStats> = Vec::new();
    let mut unused_tags: Vec<String> = Vec::new();

    let registered: Vec<String> = match tags_data {
        Some(td) => td
            .categories
            .iter()
            .flat_map(|c| c.tags.iter().map(|t| t.name.clone()))
            .collect(),
        None => Vec::new(),
    };

    if tags_file_found {
        if let Some(td) = tags_data {
            for cat in &td.categories {
                let mut category_tags: Vec<TagCount> = Vec::new();
                for tag_def in &cat.tags {
                    let count = lookup_count(tag_counts, &tag_def.name);
                    if count > 0 {
                        category_tags.push(TagCount {
                            tag: tag_def.name.clone(),
                            count,
                        });
                    }
                }
                category_tags.sort_by_key(|t| std::cmp::Reverse(t.count));
                if !category_tags.is_empty() {
                    categories.push(CategoryStats {
                        name: cat.name.clone(),
                        tags: category_tags,
                    });
                }
            }
        }

        let mut unregistered_tags: Vec<TagCount> = tag_counts
            .iter()
            .filter(|(name, _)| !registered.iter().any(|r| r == name))
            .map(|(name, count)| TagCount {
                tag: name.clone(),
                count: *count,
            })
            .collect();
        if !unregistered_tags.is_empty() {
            unregistered_tags.sort_by_key(|t| std::cmp::Reverse(t.count));
            categories.push(CategoryStats {
                name: "Unregistered".to_string(),
                tags: unregistered_tags,
            });
        }

        if let Some(td) = tags_data {
            for cat in &td.categories {
                for tag_def in &cat.tags {
                    if lookup_count(tag_counts, &tag_def.name) == 0 {
                        unused_tags.push(tag_def.name.clone());
                    }
                }
            }
            unused_tags.sort();
        }
    } else if !tag_counts.is_empty() {
        let mut all_tags: Vec<TagCount> = tag_counts
            .iter()
            .map(|(name, count)| TagCount {
                tag: name.clone(),
                count: *count,
            })
            .collect();
        all_tags.sort_by_key(|t| std::cmp::Reverse(t.count));
        categories.push(CategoryStats {
            name: "Unregistered".to_string(),
            tags: all_tags,
        });
    }

    (categories, unused_tags)
}

fn lookup_count(tag_counts: &[(String, usize)], tag: &str) -> usize {
    tag_counts
        .iter()
        .find(|(k, _)| k == tag)
        .map(|(_, c)| *c)
        .unwrap_or(0)
}

fn render_text(result: &TagStatsResult) -> String {
    let mut out = String::new();
    out.push('\n');
    out.push_str("Tag Usage Statistics\n");
    out.push_str(&"\u{2500}".repeat(50));
    out.push('\n');
    out.push_str(&format!("Total feature files: {}\n", result.total_files));
    out.push_str(&format!("Unique tags used: {}\n", result.unique_tags));
    out.push_str(&format!(
        "Total tag occurrences: {}\n",
        result.total_occurrences
    ));

    if !result.tags_file_found {
        out.push('\n');
        out.push_str("\u{26A0} Warning: spec/tags.json not found\n");
    }

    if !result.invalid_files.is_empty() {
        out.push('\n');
        out.push_str(&format!(
            "\u{26A0} Warning: {} file(s) with invalid syntax skipped:\n",
            result.invalid_files.len()
        ));
        for f in &result.invalid_files {
            out.push_str(&format!("  - {f}\n"));
        }
    }

    if !result.categories.is_empty() {
        out.push_str("\n\n");
        out.push_str("Tag Counts by Category\n");
        out.push_str(&"\u{2500}".repeat(50));
        out.push('\n');

        for category in &result.categories {
            out.push('\n');
            out.push_str(&format!(
                "{} ({} tags)\n",
                category.name,
                category.tags.len()
            ));
            for tc in &category.tags {
                let padded = pad_end(&tc.tag, 30);
                out.push_str(&format!("  {padded} {}\n", tc.count));
            }
        }
    }

    if !result.unused_tags.is_empty() {
        out.push_str("\n\n");
        out.push_str("Unused Registered Tags\n");
        out.push_str(&"\u{2500}".repeat(50));
        out.push('\n');
        out.push_str(&format!(
            "{} registered tag(s) not used in any feature file:\n\n",
            result.unused_tags.len()
        ));
        for tag in &result.unused_tags {
            out.push_str(&format!("  {tag}\n"));
        }
    }

    out.push('\n');
    out
}

fn pad_end(s: &str, width: usize) -> String {
    let len = s.chars().count();
    if len >= width {
        s.to_string()
    } else {
        let pad = width - len;
        let mut out = String::with_capacity(s.len() + pad);
        out.push_str(s);
        for _ in 0..pad {
            out.push(' ');
        }
        out
    }
}

fn parse_feature_tags(content: &str) -> Option<Vec<String>> {
    let mut pending_tags: Vec<String> = Vec::new();
    let mut feature_tags: Option<Vec<String>> = None;

    for raw in content.lines() {
        let trimmed = raw.trim();
        if trimmed.is_empty() {
            continue;
        }
        if trimmed.starts_with('#') {
            continue;
        }
        if trimmed.starts_with('@') {
            for tok in trimmed.split_whitespace() {
                if tok.starts_with('@') {
                    pending_tags.push(tok.to_string());
                }
            }
            continue;
        }
        if trimmed.starts_with("Feature:") {
            if feature_tags.is_none() {
                feature_tags = Some(std::mem::take(&mut pending_tags));
            } else {
                pending_tags.clear();
            }
            continue;
        }
        pending_tags.clear();
    }

    feature_tags
}

#[cfg(test)]
mod tests {
    #![allow(
        clippy::unwrap_used,
        clippy::expect_used,
        clippy::panic,
        clippy::useless_vec
    )]
    use super::*;

    #[test]
    fn pad_end_pads_short_strings_with_spaces() {
        assert_eq!(pad_end("abc", 5), "abc  ");
        assert_eq!(pad_end("abc", 3), "abc");
        assert_eq!(pad_end("longer", 3), "longer");
    }

    #[test]
    fn bump_count_preserves_insertion_order() {
        let mut counts: Vec<(String, usize)> = Vec::new();
        bump_count(&mut counts, "@b");
        bump_count(&mut counts, "@a");
        bump_count(&mut counts, "@b");
        assert_eq!(counts[0].0, "@b");
        assert_eq!(counts[0].1, 2);
        assert_eq!(counts[1].0, "@a");
        assert_eq!(counts[1].1, 1);
    }

    #[test]
    fn parse_feature_tags_extracts_feature_level_only() {
        let body = "@critical\nFeature: X\n\n  @smoke\n  Scenario: A\n    Given x\n";
        let tags = parse_feature_tags(body).expect("Some for valid Feature");
        assert_eq!(tags, vec!["@critical".to_string()]);
    }

    #[test]
    fn parse_feature_tags_none_when_no_feature_header() {
        assert!(parse_feature_tags("This is not gherkin at all\n").is_none());
    }
}
